src/main/scala/viper/gobra/translator/encodings/structs/StructEncoding.scala

```diff
@@ -167,35 +222,97 @@ class StructEncoding extends TypeEncoding {
+        vVal <- ctx.expression(upd.newVal)
+      } yield pex.update(vBase, idx, vVal, cptParam(fs)(ctx))(upd)(ctx)
+
+    case (e: in.DfltVal) :: ctx.CompleteStruct(fs) / Exclusive =>
       val fieldDefaults = fs.map(f => in.DfltVal(f.typ)(e.info))
       sequence(fieldDefaults.map(ctx.expression)).map(ex.create(_, cptParam(fs)(ctx))(e)(ctx))
 
+    case (e: in.DfltVal) :: ctx.PartialStruct(fs) / Exclusive =>
+      val fieldDefaults = fs.map(f => in.DfltVal(f.typ)(e.info))
+      sequence(fieldDefaults.map(ctx.expression)).map(pex.create(_, cptParam(fs)(ctx))(e)(ctx)) 
+
     case (e: in.DfltVal) :: ctx.Struct(fs) / Shared =>
       val (pos, info, errT) = e.vprMeta
       unit(shDfltFunc(Vector.empty, fs)(pos, info, errT)(ctx))
 
-    case (lit: in.StructLit) :: ctx.Struct(fs) =>
+    case (lit: in.StructLit) :: ctx.CompleteStruct(fs) =>
       val fieldExprs = lit.args.map(arg => ctx.expression(arg))
+      println(s"complete lit: $lit, ${ctx.lookupConstructor(lit.typ)}")
       sequence(fieldExprs).map(ex.create(_, cptParam(fs)(ctx))(lit)(ctx))
+      /* val construct = ctx.lookupConstructor(lit.typ)
+      if (construct.isEmpty) { sequence(fieldExprs).map(ex.create(_, cptParam(fs)(ctx))(lit)(ctx)) }
+      else {
+        val (pos, info, errT) = lit.vprMeta
+        println(s"complete lit2: $lit, ${ctx.lookupConstructor(lit.typ)}")
+        val funcName = construct.getOrElse(null).uniqueName
+        val src = lit.info
+        val z = in.LocalVar(ctx.freshNames.next(), lit.typ)(src)
+        val v = variable(ctx)(z).localVar
+        write(vpr.FuncApp(funcName, Seq(v))(pos, info, v.typ, errT))
+      } */
+
+    case (lit: in.StructLit) :: ctx.PartialStruct(fs) =>
+      val fieldExprs = lit.args.map(arg => ctx.expression(arg)).zip(fs.map(_.notImported)).collect { case (e, false) => e }
+      sequence(fieldExprs).map(pex.create(_, cptParam(fs)(ctx))(lit)(ctx)) 
 
-    case (loc: in.Location) :: ctx.Struct(_) / Shared =>
-      sh.convertToExclusive(loc)(ctx, ex)
+    case (loc: in.Location) :: ctx.CompleteStruct(_) / Shared =>
+      val deref = ctx.lookupDereference(loc.typ)
+      if (deref.isEmpty) { sh.convertToExclusive(loc)(ctx, ex) }
+      else { 
+        val (pos, info, errT) = loc.vprMeta
+        val funcName = deref.getOrElse(null).uniqueName
+        for {
+          b <- bind(loc)(ctx)
+          v = variable(ctx)(b).localVar
+          func = vpr.FuncApp(funcName, Seq(v))(pos, info, v.typ, errT)
+        } yield func
+      }
+    
+    case (loc: in.Location) :: ctx.PartialStruct(_) / Shared =>
+      val deref = ctx.lookupDereference(loc.typ)
+      if (deref.isEmpty) { sh.convertToExclusive(loc)(ctx, pex) }
+      else { 
+        val (pos, info, errT) = loc.vprMeta
+        val funcName = deref.getOrElse(null).uniqueName
+        for {
+          b <- bind(loc)(ctx)
+          v = variable(ctx)(b).localVar
+          func = vpr.FuncApp(funcName, Seq(v))(pos, info, v.typ, errT)
+        } yield func
+      }
```

Comment:
Why should we throw an error if there is no deref function? Since there is a default constructor if there is no constructor, then we can just use the default dereference (`convertToExclusive`) that just gives back all the values that it has permission to and is visible.

Reply:
Additionally, if we always use `sh.convertToExclusive(...)` for a CompleteStruct, then this statement would fail: 
```
requires acc(&a.Left) && acc(&a.Right) && a.PrivateMem()
func foo(a *Pair) {
  x := *a // fails, not enough permission
}
```
Since there is no call to dereference, `Pair `that has fields `Left`, `Right`, `p` does not have access to the field `p` (within the same package until we `unfold a.PrivateMem()`). The constructor always gets called once defined, therefore:
```
x := &Pair{Left: 2, Right: 0, p: 1}
assert(x.PrivateMem()) //constructor ensures x.PrivateMem()
```
and this assignment would fail, despite constructing `x` just before: `y := *x`

The question is more if we want to explicitly unfold x.PrivateMem() or just let the dereference handle it.